docs/KFSERVING_DEBUG_GUIDE.md

```diff
@@ -0,0 +1,335 @@
+    route:
+    - destination:
+        host: cluster-local-gateway.istio-system.svc.cluster.local
+        port:
+          number: 80
+      headers:
+        request:
+          set:
+            Host: sklearn-iris-predictor-default.default.svc.cluster.local
+      weight: 100
+```
+- KFServing creates the routing rule based on uri prefix according to [KFServing V1 data plane](./README.md#data-plane-v1) and traffic
+is forwarded to `KFServing Predictor` if you only have `Predictor` specified on `InferenceService`,
+note that if you have custom container and the endpoint is not conforming to the protocol you get `HTTP 404` when you hit the KFServing
+top level virtual service.
+- When `Transformer` and `Explainer` are specified on `InferenceService` the routing rule sends the traffic to `Transformer`
+or `Explainer` based on the verb.
+- The top level virtual service also does `Canary Traffic Split` if canary is specified on `InferenceService`.
+
+3. KNative creates a `Istio virtual service` to configure the gateway to route the user traffic to correct revision
+The request then hits `Knative` created virtual service via local gateway, it matches with the in cluster host name.
+```bash
+kubectl get vs sklearn-iris-predictor-default -oyaml
+```
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: sklearn-iris-predictor-default-mesh
+  namespace: default
+spec:
+  gateways:
+  - mesh
+  hosts:
+  - sklearn-iris-predictor-default.default
+  - sklearn-iris-predictor-default.default.svc
+  - sklearn-iris-predictor-default.default.svc.cluster.local
+  http:
+  - headers:
+      request:
+        set:
+          K-Network-Hash: dee002f4a2db24e3827d8088b7ddacf3
+    match:
+    - authority:
+        prefix: sklearn-iris-predictor-default.default
+      gateways:
+      - mesh
+    retries:
+      attempts: 3
+      perTryTimeout: 600s
+      retryOn: 5xx,connect-failure,refused-stream,cancelled,resource-exhausted,retriable-status-codes
+    route:
+    - destination:
+        host: sklearn-iris-predictor-default-fhmjk.default.svc.cluster.local
+        port:
+          number: 80
+      headers:
+        request:
+          set:
+            Knative-Serving-Namespace: default
+            Knative-Serving-Revision: sklearn-iris-predictor-default-fhmjk
+      weight: 100
+    timeout: 600s
+    websocketUpgrade: true
+```
+The destination here is the `k8s Service` for the latest ready `Knative Revision` and it is reconciled by `Knative` every time
+user rolls out a new revision. When a new revision is rolled out and in ready state, the old revision is then scaled down, after
+configured revision GC time the revision resource is garbage collected if the revision no longer has traffic referenced.
+
+4. Once the revision pods are ready, the `Kubernetes Service` sends the requests to the `queue proxy` sidecar on `port 8012`.
+```bash
+kubectl get svc sklearn-iris-predictor-default-fhmjk-private -oyaml
```

Comment:
It might be worth mentioning that this 'private service' can be tested by 
`kubectl port-forward svc/sklearn-iris-predictor-default-fhmjk-private 80:8080`
and
`curl -v -H "Content-Type: application/json" http://localhost:8080/v1/models/sklearn-iris:predict -d @iris-input.json`